vignettes/custom-functions.Rmd

```diff
@@ -0,0 +1,109 @@
+
+## Writing functions using `polars` expressions
+
+One of `polars`' biggest strength is the composability of expressions.
+Chaining multiple functions to create custom functions is quite simple.
+For example, `polars` doesn't provide a built-in function to standardize a variable, but we could create one using expressions:
+
+```{r}
+pl_standardize <- function(x) {
+  (x - x$mean()) / x$std()
+}
+```
+
+And we can then run it in `$with_columns()` for instance:
+```{r}
+dat <- as_polars_df(mtcars[, c("carb", "mpg", "drat")])
+dat$with_columns(
+  carb_stand = pl_standardize(pl$col("carb"))
+)
+```
+
+## Using `purrr`
+
+`polars` expressions allow one to write their own functions, but it is sometimes necessary to use functions that already exist in other R packages.
+When this is the case, one should keep in mind that applying functions that are external to `polars` will limit `polars`' strengths.
+In particular, `polars` is fast in part because it is able to detect the various transformations applied to the data and run them in parallel or cache them so that only the strict minimum of computations is made.
+Using `polars` expressions will also allow the code to be run in streaming mode, allowing for larger-than-memory data.
+When one uses functions that `polars` doesn't natively know, those benefits are lost.
+This is the reason why it is better to use `polars` expressions in your own functions if you can.
+
+To apply an external function to a `polars` object, we must convert this object to R, apply the function, and then convert it back to `polars`.
+If you come from Python Polars, this is a big difference.
+In Python Polars, several functions (for instance `map_batches()` and `map_elements()`) allow one to pass custom functions.
+Those are not in the R implementation because we cannot guarantee that they will function correctly in streaming mode.
+
+Note that when you convert your `polars` object to R, you may run out of memory if your object is a LazyFrame that is too big to fit in memory once collected.
+Another crucial point is that we want the R function to run as fast as possible on the data.
+One way to do that is to take advantage of `purrr` to run the function [in parallel](https://purrr.tidyverse.org/dev/reference/parallelization.html).
+
+**Note:** at the time of writing, this requires the development version of `purrr` (1.0.4.9000) and the package `mirai`.
+
+```{r}
+# Set the number of background processes to use
+mirai::daemons(3)
+
+dat |>
+  as.list(as_series = FALSE) |>
+  purrr::map(\(s) scale(s)[, 1], .parallel = TRUE) |>
+  _[] |>
+  as_polars_df()
+```
+
+Note that this applies the function on the entire DataFrame (or LazyFrame).
+To apply this to a subset of columns only, we can use `map_at()`:
+
+```{r}
+dat |>
+  as.list(as_series = FALSE) |>
+  purrr::map_at(c("carb", "mpg"), \(s) scale(s)[, 1], .parallel = TRUE) |>
+  _[] |>
+  as_polars_df()
+```
+
+
+## Using `mirai`
```

Comment:
This section is a bit empty, but I don't know what I can add. I added a comment below about running LazyFrame on a remote machine, but I have no idea how that would work (or if it even does).

If we do talk about this feature, maybe we should mention this is not the same as Polars Cloud.